content/blog/2019-10-23-AutoVectoring/index.md

```diff
@@ -0,0 +1,242 @@
+
+After we are confident that the array operations in a loop are vectorizable, we now need to convert the loop structure to its equivalent vectorized form.
+
+### Strip Mining
+Since we are working with vector operations on four consecutive array elements, we "chunk" the sequence of loop iterations in blocks of four, which is known as *strip mining*. 
+
+Trivially, strip mining can be done by finding the statement that increments the induction variable and changing it from incrementing/decrementing 1 to incrementing/decrementing 4. All operation on array elements are changed to their vector counterparts, e.g., `add` would become `vadd`. This would work as long as the loop only included operations on arrays, but we found that to often not be the case. 
+
+For example, printing the *i* each iteration, where *i* is the induction variable, should still allow the loop to be vectorized as there are no flow-dependencies, but if we change the induction variable by four each iteration, that print statment will behave incorrectly as it would only print one out of four elements. To mitigate this, we do partial loop unrolling when strip mining.
+
+To achieve this, we go through the loop and find non-array instructions and keep track of them. Then, we append them to the end of the loop (before the condition and branch statements) and also a copy of the induction variable increment/decrement. We do this insertion three times total. 
+
+Example loop snippit before strip mining and partial unrolling:
+```gherkin=
+loop:
+  print i;
+  ci: int = add c i;
+  v: int = lw ci;
+  i: int = add i inc;
+  done: bool = eq i size;
+  br done exit loop;
+```
+After strip mining and partial unrolling:
+```gherkin=
+loop:
+  print i;
+  ci: int = add c i;
+  v: int = vload ci;
+  i: int = add i inc;
+  print i;
+  i: int = add i inc;
+  print i;
+  i: int = add i inc;
+  print i;
+  i: int = add i inc;
+  done: bool = eq i size;
+  br done exit loop;
+```
+
+With this method, array operations are allowed to be vectorized while preserving serial instructions because serial instructions are unrolled into four copies, where each copy operates with a different induction variable value. This loop also increments the induction variable by four every iteration which preserves the performance increase from reduced branch overhead. 
+
+In terms of implentaion, we found it much easier to first coalesce the loop into one big block before running the strip mine algorithm because we could treat this sequence of instructions as one array instead of having to worry about jumps and label renaming from inserting new blocks. Aggregating all the blocks of this loop was possible because we previously enforced that there can be exactly one branch instruction located at the end of the loop.
```

Comment:
```suggestion
In terms of implementation, we found it much easier to first coalesce the loop into one big block before running the strip mine algorithm because we could treat this sequence of instructions as one array instead of having to worry about jumps and label renaming from inserting new blocks. Aggregating all the blocks of this loop was possible because we previously enforced that there can be exactly one branch instruction located at the end of the loop.
```